### ISSUE TITLE
NeighbourLists

### ISSUE THREAD

**Issue by cortner** [m3g/CellListMap.jl#3]
discuss cell list imlementation here vs [NeighbourLists.jl](https://github.com/JuliaMolSim/NeighbourLists.jl) - worth combinig code an efforts?

**Comment by lmiq** [comment]
NeighbourLists:   1.091 s (800095 allocations: 293.98 MiB)

this kind of allocations suggest that there is some type instability in the code, which may be causing most of the slowdown. 

Apart from that, I have tweaked quite a few things. I run only over the cells which are "forward" on the list to avoid repetitions (thus over 13 cells per cell, not 26). The cell size is an [adjustable parameter ](https://github.com/m3g/CellListMap.jl#optimizing-the-cell-grid), but by default this parameter is 1 and the cell size is the same as the cutoff. There are many other small optimizations that I have done in the last few weeks also. 

I will effectively implement support for general periodic boundary conditions, I expect very soon, this is on my priority list. Also, I will try to implement the ideas on this paper as well: https://www.google.com/url?sa=t&rct=j&q=&esrc=s&source=web&cd=&cad=rja&uact=8&ved=2ahUKEwiClvCWqLrxAhWMpJUCHZZ8D-wQFjABegQIAxAD&url=https%3A%2F%2Farxiv.org%2Fabs%2F1804.06231&usg=AOvVaw2m5jTOkSYfeTx5pLaehCar

But my goal is not to provide the best strategy for MD simulations. Actually for MD simulations the best approach, the Verlet Lists, is easier: build a list of neighbours in the first step (maybe using cell lists), and update these lists after some steps only for the particles that moved outside the boundaries. This is much faster than the cell lists approach, if you can guarantee that there is correlation between the positions of the particles among successive frames. This is not my case though, as for analysis routines the frames are to far from each other in time, and for Packmol, there might be no correlation between the positions of the particles in different steps of the optimization procedure.

**Comment by cortner** [comment]
We never bothered with verlet lists since we are interested in potentials that are vastly more expensive than a neighbourlist assembly.

**Comment by cortner** [comment]
The profiling doesn’t show an allocation where the bottleneck is. I thought this meant there is no type instability. Otherwise I 100% agree it looks just like that

**Comment by lmiq** [comment]
That is strange. Maybe a good pass of `@code_warntype` in different functions finds something.

**Comment by cortner** [comment]
I will definitely do that next once I find the time :). Atm we have school closure due to heatwave so no chance…

**Comment by lmiq** [comment]
@cortner, just to inform that version 0.5.1 now supports arbitrary periodic boundary conditions (and should be somewhat faster yet, particularly for dense systems).

**Comment by cortner** [comment]
that's really great, thanks for mentioning it. I'll run it through my test set and maybe I can retire my NeighbourLists.jl package.

**Comment by lmiq** [comment]
I am working actively on the package, and it will become faster still (it will fall back to different strategies of computation depending on system size and density). I still have to set a scheme for if someone does not want periodic boundary conditions at all, but that will be easy.

**Comment by m3g** [comment]
Not sure if you had time to test anything, but I noticed that the general PBCs where failing in many cases. Now I uploaded a new version, 0.5.2, which will raise errors if the cell cannot be handled. In particular I can only handle correctly cells with all positive lattice vectors and which are not collapsed relative to the cutoff. At some point I will try to follow the same rules either Gromacs or Lammps follow, but that is much more cumbersome than I initially imagined.

**Comment by cortner** [comment]
Sorry I didn’t get to it yet, too many other things going on. But I will for sure look into it.

**Comment by lmiq** [comment]
The package provides a `neighbourlist` function that implements this calculation. Without periodic boundary conditions, just do:

```julia-repl
julia> x = [ rand(3) for _ in 1:10_000 ];

julia> CellListMap.neighbourlist(x,0.05)
24778-element Vector{Tuple{Int64, Int64, Float64}}:
 (1, 62, 0.028481068525796384)
 ⋮
 (9954, 1749, 0.04887502372299809)
 (9974, 124, 0.040110356034451795)
```
or `CellListMap.neighbourlist(x,y,r)` for computing the lists of pairs of two sets closer than `r`.

The returning array contains tuples with the index of the particle in the first vector, the index of the particle in the second vector, and their distance.

If periodic boundary conditions are used, the `Box` and `CellList` must be constructed in advance:
```julia-repl
julia> x = [ rand(3) for _ in 1:10_000 ]; 

julia> box = Box([1,1,1],0.1);

julia> cl = CellList(x,box);

julia> CellListMap.neighbourlist(box,cl)

julia> CellListMap.neighbourlist(box,cl)
209506-element Vector{Tuple{Int64, Int64, Float64}}:
 (1, 121, 0.05553035041478053)
 (1, 1589, 0.051415489701932444)
 ⋮
 (7469, 7946, 0.09760096646331885)
```

(this is part of the documentation now, and will be released exactly like this in v0.7)
